### DISCUSSION TITLE
Landuse polygon merging

### DISCUSSION THREAD

**Discussion by adjourn** [protomaps/basemaps#322] [General]
Experimenting with different planetiler profiles out there..

I love the simplicity and straightforwardness of this profile, operating directly on planetiler constructs instead of building a custom layer or few between. I also love most defaults, when water labels are shown, how POIs are handled on different zoom levels instead of just throwing everything into z14, etc etc. Very good stuff!

However it has a weight problem in my opinion, **specifically z7-9** which seem to be 2-2.5x heavier than [planetiler-openmaptiles](https://github.com/openmaptiles/planetiler-openmaptiles).
It seems to come down to not merging the landuse polygons on these zoom levels.

Any other reason besides "haven't got to it yet"?

---

Picked a random Europe country tile on z7: 
* protomaps 362kB (used [this map](https://maps.protomaps.com/#map=0.78/0/74.6&theme=light&lang=en&tiles=https://demo-bucket.protomaps.com/v4.pmtiles&show_boxes=true))
* openmaptiles 127kB (my generated, has admin 6 borders as well, otherwise looks identical details wise)

**Comment by bdon** [comment]
simply haven't gotten to it yet, though @nvkelso has devoted a lot to optimizing this in Tilezen: https://tilezen.readthedocs.io/en/latest/layers/#low-zoom-consolidation

**Reply by adjourn** [reply]
I'll maybe then contribute if I decide to land on `protomaps` profile.

I've written Java only for a month when on this map jurney 🙃, day job with other porgramming languages..
But it seems all the pieces are there, just have to put them together.

```java
// Landuse.java 

import com.onthegomap.planetiler.FeatureMerge;

// in post process.. 
@Override
public List<VectorTile.Feature> postProcess(int zoom, List<VectorTile.Feature> items) throws GeometryException {
  List<VectorTile.Feature> list = new ArrayList<>();
  List<VectorTile.Feature> output = new ArrayList<>();
  
  for (var item : items) {
      if (/* some kind of condition what to merge */) {
        list.add(item);
      } else {
        // if some features should not be merged..
        output.add(item);
      }
    }

  // these should be dependent on zoom level
  double dist = 0.1;
  double buffer = 0.1;
  // merge them (for example only in certail zoom levels)
  List<VectorTile.Feature> merged = FeatureMerge.mergeNearbyPolygons(list, 1, 1, dist, buffer);

  output.addAll(merged);
  return output;
}
```

Why all features have `$id` attribute? Asking because merged polygons can't have any unique fields.

**Reply by adjourn** [reply]
Just noticed you already have similar logic, but only `if (zoom <= 6)..` for some reason.
And since you do this `.setId(FeatureId.create(sf))`, those can't be merged because each polygon has an unique attribute.

**Reply by bdon** [reply]
The `mergeNearbyPolygons` should work by merging the same tags, ignoring IDs. setting an ID is required for interaction features in maplibre GL like feature states / hovering.

**Reply by adjourn** [reply]
I think you're right, it ignores IDs on merge.

Did a quick experiment.
This change alone:
```java
if (zoom <= 7) { // <- was 6
  return FeatureMerge.mergeNearbyPolygons(items, 3.125, 3.125, 0.5, 0.5);
}
```

Zoom 7
Random Europe tile with lots of forest, farmland, etc (temperate climate)
* Was 362kB
* Now 120kB

Can be optimized even further by e.g combining `forest` and `wood`, if the distinction is not necessary.
But yeah, it's a question of where this profile stands.. Performance by default or interactivity (detailed features).

**Reply by nvkelso** [reply]
Great initial results with the simple solution, woot! This approach is generic, we used it in Tilezen as well.

Care to file a PR for the basic use case, and then if we need to later we can do the grouping of kinds pre-merge in followup PR. Clearly the basic PR will have huge impact and we should get that in 1st :)
